sql/v2/expression/like_expression.go

```diff
@@ -30,70 +27,61 @@ func (l likeExpression) Evaluate(event cloudevents.Event) (interface{}, error) {
 		return nil, err
 	}
 
-	return l.pattern.MatchString(val.(string)), nil
+	return matchString(val.(string), l.pattern), nil
+
 }
 
 func NewLikeExpression(child cesql.Expression, pattern string) (cesql.Expression, error) {
-	// Converting to regex is not the most performant impl, but it works
-	p, err := convertLikePatternToRegex(pattern)
-	if err != nil {
-		return nil, err
-	}
-
 	return likeExpression{
 		baseUnaryExpression: baseUnaryExpression{
 			child: child,
 		},
-		pattern: p,
+		pattern: pattern,
 	}, nil
 }
 
-func convertLikePatternToRegex(pattern string) (*regexp.Regexp, error) {
-	var chunks []string
-	chunks = append(chunks, "^")
-
-	var chunk strings.Builder
+func matchString(text, pattern string) bool {
+	textLen := len(text)
+	patternLen := len(pattern)
+	textIdx := 0
+	patternIdx := 0
+	lastWildcardIdx := -1
+	lastMatchIdx := 0
 
-	for i := 0; i < len(pattern); i++ {
-		if pattern[i] == '\\' && i < len(pattern)-1 {
-			if pattern[i+1] == '%' {
-				// \% case
-				chunk.WriteRune('%')
-				chunks = append(chunks, "\\Q"+chunk.String()+"\\E")
-				chunk.Reset()
-				i++
-				continue
-			} else if pattern[i+1] == '_' {
-				// \_ case
-				chunk.WriteRune('_')
-				chunks = append(chunks, "\\Q"+chunk.String()+"\\E")
-				chunk.Reset()
-				i++
-				continue
-			} else {
-				// if there is an actual literal \ character, we need to include that in the string
-				chunk.WriteRune('\\')
-			}
-		} else if pattern[i] == '_' {
-			// replace with .
-			chunks = append(chunks, "\\Q"+chunk.String()+"\\E")
-			chunk.Reset()
-			chunks = append(chunks, ".")
-		} else if pattern[i] == '%' {
-			// replace with .*
-			chunks = append(chunks, "\\Q"+chunk.String()+"\\E")
-			chunk.Reset()
-			chunks = append(chunks, ".*")
+	for textIdx < textLen {
+		// handle escaped characters -> pattern needs to increment two places here
+		if patternIdx < patternLen-1 && pattern[patternIdx] == '\\' &&
+			((pattern[patternIdx+1] == '_' || pattern[patternIdx+1] == '%') &&
+				pattern[patternIdx+1] == text[textIdx]) {
+			patternIdx += 2
+			textIdx += 1
+			// handle non escaped characters
+		} else if patternIdx < patternLen && (pattern[patternIdx] == '_' || pattern[patternIdx] == text[textIdx]) {
+			textIdx += 1
+			patternIdx += 1
+			// handle wildcard characters
+		} else if patternIdx < patternLen && pattern[patternIdx] == '%' {
+			lastWildcardIdx = patternIdx
+			lastMatchIdx = textIdx
+			patternIdx += 1
+			// greedy match didn't work, try again from the last known match
+		} else if lastWildcardIdx != -1 {
```

Comment:
I'm probably wrong, and it's probably due to the nature of % being so greedy, but are we sure that we don't need a stack of "last known wildcard indexes" instead of just 1? I'm wondering about a case where we've seen multiple %'s  and then we don't match something and we need to back-up to the first % not just the 2nd %. I can't seem to write a testcase to show the issue, hence why I'm probably wrong, but I thought I'd mention it/ask....

Reply:
For me the trick to understanding this is that wildcard matching has optimal substructure: if we have already matched part of the input string against part of the pattern string, we only need to use the remaining parts of the pattern string and input string to determine if the entire string matches the pattern or not.

The only slight "exception" to the optimal substructure is that the `%` character can consume more than one element of the input. So, if at some point our pattern no longer matches the input text, we need to go back to the last wildcard character we saw, as well as the place we were at in the input text when we saw it and try again, consuming one character from the input text in the process.

The key is that we _know_ that we have a valid match up until the last wildcard character we saw, and that apart from going back to that wildcard and consuming more characters from the input text there we can not otherwise move backwards in the pattern or the input text. So, storing an older index would not end up being used, since it would be in part of the input text we had already matched.